### PR TITLE
[Feature] Flow through the iceberg table sink

### PR DESCRIPTION
**Pull request by stephen-shelby**
## What type of PR is this：
- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #
https://github.com/StarRocks/starrocks/issues/21502

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto backported to target branch
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 22 / 52 (42.31%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/planner/IcebergTableSink.java|0|17|00.00%|[67, 68, 69, 70, 71, 72, 77, 78, 79, 80, 81, 82, 83, 84, 85, 87, 88]|
|:large_blue_circle:|com/starrocks/catalog/IcebergTable.java|0|1|00.00%|[213]|
|:large_blue_circle:|com/starrocks/qe/Coordinator.java|2|4|50.00%|[423, 1493]|
|:large_blue_circle:|com/starrocks/planner/DataSink.java|1|2|50.00%|[97]|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|19|28|67.86%|[1494, 1550, 1551, 1552, 1553, 1554, 1557, 1558, 1559]|

File: be/src/exec/parquet_writer.h [hunk 1]
```diff
@@ -39,62 +39,48 @@ struct TableInfo;
...
     Status close_current_writer(RuntimeState* state);
 
+private:
     std::shared_ptr<FileSystem> _fs;
     std::shared_ptr<starrocks::parquet::AsyncFileWriter> _writer;
     std::shared_ptr<::parquet::WriterProperties> _properties;
     std::shared_ptr<::parquet::schema::GroupNode> _schema;
-    std::string _partition_dir;
+    std::string _partition_location;
     int32_t _cnt = 0;
```
Comment (letian-jiang) [inline be/src/exec/parquet_writer.h:77]: better to use `_outfile_cnt`

File: be/src/runtime/descriptors.h [hunk 2]
```diff
@@ -210,9 +210,14 @@ class IcebergTableDescriptor : public HiveTableDescriptor {
     ~IcebergTableDescriptor() override = default;
     bool has_partition() const override { return false; }
     const TIcebergSchema* get_iceberg_schema() const { return &_t_iceberg_schema; }
+    bool is_unpartitioned_table() { return _partition_column_names.empty(); }
```
Comment (letian-jiang) [inline be/src/runtime/descriptors.h:213]: maybe override `has_paritition()` in base class

File: be/src/exec/pipeline/sink/iceberg_table_sink_operator.cpp [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+bool IcebergTableSinkOperator::pending_finish() const {
+    return !is_finished();
+}
+
+Status IcebergTableSinkOperator::set_cancelled(RuntimeState* state) {
+    return Status::OK();
+}
+
+StatusOr<ChunkPtr> IcebergTableSinkOperator::pull_chunk(RuntimeState* state) {
+    return Status::InternalError("Shouldn't pull chunk from mysql table sink operator");
```
Comment (letian-jiang) [inline be/src/exec/pipeline/sink/iceberg_table_sink_operator.cpp:84]: inconsistent operator name

File: be/src/exec/pipeline/sink/iceberg_table_sink_operator.cpp [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+        //        for (size_t i = 0; i < partitions_columns.size(); ++i) {
+        //            ASSIGN_OR_RETURN(partitions_columns[i], _partition_expr[i]->evaluate(chunk.get()));
+        //            DCHECK(partitions_columns[i] != nullptr);
+        //        }
+        //
+        //        std::vector<std::string> partition_column_names = _iceberg_table->partition_column_names();
+        //
+        //        std::vector<std::string> partition_column_values;
+        //        for (const ColumnPtr& column : partitions_columns) {
+        //            partition_column_values.emplace_back(_value_to_string(column, 0));
```
Comment (letian-jiang) [inline be/src/exec/pipeline/sink/iceberg_table_sink_operator.cpp:116]: Could we assume this chunk belongs to the same partition?

File: be/src/runtime/descriptors.h (hunk 2 above)
Comment (stephen-shelby) [reply be/src/runtime/descriptors.h:213]: if I override `has_partition()`, there will be a lot of adaptation work during iceberg scan stage.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 22 / 55 (40.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/planner/IcebergTableSink.java|0|20|00.00%|[59, 63, 65, 71, 72, 73, 74, 75, 76, 81, 82, 83, 84, 85, 86, 87, 88, 89, 91, 92]|
|:large_blue_circle:|com/starrocks/catalog/IcebergTable.java|0|1|00.00%|[213]|
|:large_blue_circle:|com/starrocks/qe/Coordinator.java|2|4|50.00%|[427, 1497]|
|:large_blue_circle:|com/starrocks/planner/DataSink.java|1|2|50.00%|[97]|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|19|28|67.86%|[1503, 1559, 1560, 1561, 1562, 1563, 1566, 1567, 1568]|

**Review by letian-jiang — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 30 / 55 (54.55%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/IcebergTable.java|0|1|00.00%|[213]|
|:large_blue_circle:|com/starrocks/planner/IcebergTableSink.java|7|20|35.00%|[63, 65, 81, 82, 83, 84, 85, 86, 87, 88, 89, 91, 92]|
|:large_blue_circle:|com/starrocks/qe/Coordinator.java|2|4|50.00%|[427, 1497]|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|19|28|67.86%|[1503, 1559, 1560, 1561, 1562, 1563, 1566, 1567, 1568]|
|:large_blue_circle:|com/starrocks/planner/DataSink.java|2|2|100.00%|[]|

**Review by letian-jiang — DISMISSED** [review]

File: be/src/exec/pipeline/pipeline_builder.cpp [hunk 5]
```diff
@@ -120,6 +120,35 @@ OpFactories PipelineBuilderContext::_maybe_interpolate_local_passthrough_exchang
     return {std::move(local_exchange_source)};
 }
 
+void PipelineBuilderContext::maybe_interpolate_local_passthrough_exchange_for_sink(RuntimeState* state,
+                                                                                   OpFactoryPtr table_sink_operator,
+                                                                                   int32_t source_operator_dop,
+                                                                                   int32_t desired_sink_dop) {
```
Comment (ZiheLiu) [inline be/src/exec/pipeline/pipeline_builder.cpp:126]: Is it possible to return directly when `source_operator_dop` is equal to `desired_sink_dop`?

File: be/src/runtime/descriptors.h [hunk 6]
```diff
@@ -210,9 +210,14 @@ class IcebergTableDescriptor : public HiveTableDescriptor {
...
+    const std::vector<std::string> partition_column_names() { return _partition_column_names; }
```
Comment (ZiheLiu) [inline be/src/runtime/descriptors.h:214]: Return `const std::vector<std::string>&`?

File: be/src/exec/pipeline/sink/iceberg_table_sink_operator.cpp [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+
+Status IcebergTableSinkOperator::push_chunk(RuntimeState* state, const ChunkPtr& chunk) {
+    TableInfo tableInfo;
+    tableInfo._schema = _parquet_file_schema;
+    tableInfo._compress_type = _compression_codec;
+
+    if (_iceberg_table->is_unpartitioned_table()) {
+        if (_partition_writers.empty()) {
+            tableInfo._partition_location = _location + "/data/";
+            auto writer = new RollingAsyncParquetWriter(tableInfo, _output_expr, _common_metrics.get(),
```
Comment (ZiheLiu) [inline be/src/exec/pipeline/sink/iceberg_table_sink_operator.cpp:95]: `unique_ptr`

File: be/src/exec/pipeline/fragment_executor.cpp [hunk 8]
```diff
@@ -911,6 +895,69 @@ Status FragmentExecutor::_decompose_data_sink_to_operator(RuntimeState* runtime_
...
+        std::vector<ExprContext*> partition_expr_ctxs;
+        auto output_expr = iceberg_table_sink->get_output_expr();
+        for (const auto& index : iceberg_table_desc->partition_index_in_schema()) {
+            partition_expr.push_back(output_expr[index]);
+        }
+
+        RETURN_IF_ERROR(Expr::create_expr_trees(runtime_state->obj_pool(), partition_expr, &partition_expr_ctxs,
+                                                runtime_state));
+        RETURN_IF_ERROR(Expr::prepare(partition_expr_ctxs, runtime_state));
+        RETURN_IF_ERROR(Expr::open(partition_expr_ctxs, runtime_state));
```
Comment (ZiheLiu) [inline be/src/exec/pipeline/fragment_executor.cpp:914]: Move to `IcebergTableSinkOperatorFactory::prepare`.

**Review by ZiheLiu — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 30 / 55 (54.55%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/IcebergTable.java|0|1|00.00%|[213]|
|:large_blue_circle:|com/starrocks/planner/IcebergTableSink.java|7|20|35.00%|[63, 65, 81, 82, 83, 84, 85, 86, 87, 88, 89, 91, 92]|
|:large_blue_circle:|com/starrocks/qe/Coordinator.java|2|4|50.00%|[427, 1497]|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|19|28|67.86%|[1503, 1559, 1560, 1561, 1562, 1563, 1566, 1567, 1568]|
|:large_blue_circle:|com/starrocks/planner/DataSink.java|2|2|100.00%|[]|

**Review by zombee0 — APPROVED** [review]

**Review by Youngwb — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 30 / 55 (54.55%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/IcebergTable.java|0|1|00.00%|[213]|
|:large_blue_circle:|com/starrocks/planner/IcebergTableSink.java|7|20|35.00%|[63, 65, 81, 82, 83, 84, 85, 86, 87, 88, 89, 91, 92]|
|:large_blue_circle:|com/starrocks/qe/Coordinator.java|2|4|50.00%|[427, 1497]|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|19|28|67.86%|[1503, 1559, 1560, 1561, 1562, 1563, 1566, 1567, 1568]|
|:large_blue_circle:|com/starrocks/planner/DataSink.java|2|2|100.00%|[]|

**Pull request merged**